packages/export/src/nodes/pivot.js

```diff
@@ -0,0 +1,139 @@
+import _ from 'lodash/fp.js'
+import F from 'futil'
+import ContextureClient from 'contexture-client'
+import { andGroup, runWith, setFilterOnly } from '../utils.js'
+
+// TODO import from contexture-client
+let getKey = (x) => x.keyAsString || x.key
+
+// TODO import from contexture-client
+// Similar to Tree.lookup but path is a drilldown which uses keyAsString or key
+let resultsForDrilldown = (type, drilldown, results) => {
+  if (_.isEmpty(drilldown) || !results) return results
+
+  let key = _.first(drilldown)
+  let groups = _.get(type, results)
+  let match = _.find((node) => getKey(node) === key, groups)
+
+  return resultsForDrilldown(type, drilldown.slice(1), match)
+}
+
+export default async ({ service, tree, exportAllPages, ...node }) => {
+  let run = (node) => runWith(service, tree, node)
+
+  let getContextureClient = (tree, node) =>
+    ContextureClient(
+      {
+        service,
+        debounce: 0,
+      },
+      andGroup(setFilterOnly(tree), node)
+    )
+
+  let cardinalityNode = await run({
+    ...node,
+    expanded: {
+      columns: true,
+      rows: true,
+      skipNested: true,
+      skipValues: true,
+      cardinality: true,
+    },
+  })
+
+  let cardinalityResult = _.get('context.results', cardinalityNode)
+
+  let getGroupingCardinality = (groupingType, result) => {
+    let getNested = _.get(groupingType)
+    let groupingCardinality = 1 // default to 1 for grand total column/row
+
+    F.walk(getNested)((groupResult, index, parents) => {
+      let groupCardinality = _.getOr(
+        _.size(getNested(groupResult)),
+        `${groupingType}Cardinality`,
+        groupResult
+      )
+      if (!groupCardinality) return
+
+      let grouping = _.get([groupingType, _.size(parents)], node)
+      if (!exportAllPages && grouping.type === 'fieldValues') {
+        groupCardinality = _.min([
+          groupCardinality,
+          _.getOr(10, 'size', grouping),
+        ])
+      }
+      // TODO else approximate full cardinality by using the size of the first page of each grouping
+
+      groupingCardinality = groupingCardinality + groupCardinality
+    })(result)
+
+    return groupingCardinality
+  }
+
+  let columnCardinality = getGroupingCardinality('columns', cardinalityResult)
+  let rowCardinality = getGroupingCardinality('rows', cardinalityResult)
+
+  let pivotCardinality =
+    columnCardinality * rowCardinality * (_.size(node.values) || 1)
+
+  let pivot = {
+    node: cardinalityNode,
+    getTotalRecords() {
+      return pivotCardinality
+    },
+    async *[Symbol.asyncIterator]() {
```

Comment:
I know you were copying other code but this is probably easier to read:

```javascript
async function* pivot() {
   ....
}

pivot.node = ...

pivot.getTotalRecords = ...

return pivot
```

Reply:
Let's stay consistent with the rest other ones for now - we're probably more likely to change the structure to `{ getTotalRecords, async* run }` or something like that than to dig deeper in this structure.